modules/authorization-xacml/src/main/java/org/opencastproject/authorization/xacml/XACMLAuthorizationService.java

```diff
@@ -308,8 +317,18 @@ private Optional<AccessControlList> loadAcl(final URI uri) {
   @Override
   public boolean hasPermission(final MediaPackage mp, final String action) {
     AccessControlList acl = getActiveAcl(mp).getA();
-    boolean allowed = false;
     final User user = securityService.getUser();
+    var allowed = false;
+
+    // Check special ROLE_EPISODE_<ID>_<ACTION> permissions
```

Comment:
The check only iterates over the user roles once, instead of over all user roles for every ACE like usual, so I'd wager there should not be much of a performance difference. But I honestly have no idea, would have to run benchmarks.

Reply:
Yeah, no hard feelings here, feel free to skip.